runtime/compiler/runtime/RuntimeAssumptions.cpp

```diff
@@ -85,10 +85,10 @@ TR_PatchNOPedGuardSiteOnClassPreInitialize::hashCode(char *sig, uint32_t sigLen)
 void
 TR_PatchNOPedGuardSiteOnClassPreInitialize::reclaim()
    {
-   TR_ASSERT_FATAL(_key != NULL, "Attempt to reclaim an already freed _key");
+   TR_ASSERT_FATAL((void*)_key != NULL, "Attempt to reclaim an already freed _key");
```

Comment:
```suggestion
   TR_ASSERT_FATAL(reinterpret_cast<void *>(_key) != NULL, "Attempt to reclaim an already freed _key");
```

Reply:
I suggest that neither of those changes are appropriate. The field `_key` is declared with type `uintptr_t` (a scalar type): `0` is the right value for a freed key.

Reply:
The ISO C++ standard defers to the ISO C standard, which gives no guarantees on the behavior of (u)intptr_t except that pointers to void may be converted to and from it. It is not guaranteed that all-zero pointers retain the same value when converted to (u)intptr_t. Furthermore, it is not guaranteed that null pointers are represented as all zeroes in the first place. However, as a practical matter, current implementations directly copy the byte representation as an integer, allow integer operations on them, represent null pointers as all zeroes, and are unlikely to change any of those behaviors or make optimization decisions based on them.

In conclusion, the cast is more correct, but any of these three alternatives are likely to be equivalent in practice.

Reply:
If `(0 != (uintptr_t)NULL)` then we'll have lots of broken code.
I still favour `0`, for consistency with other places that use `uintptr_t`.

Reply:
yeah, I think we can ignore `(0 != (uintptr_t)NULL)` for now. it would break a huge amount of existing code in other projects too, and AFAIK doesn't have any good optimization rationale. @dsouzai are you OK with `0 != _key` here?

Reply:
Yeah no issues.